test/integration/jobset_controller_test.go

```diff
@@ -0,0 +1,264 @@
+				Type:   batchv1.JobFailed,
+				Status: corev1.ConditionTrue,
+			})
+			Expect(k8sClient.Status().Update(ctx, &job)).Should(Succeed())
+
+			// Check JobSet has failed.
+			Eventually(assertJobSetStatus, timeout, interval).WithArguments(js, jobset.JobSetFailed).Should(Equal(true))
+		})
+	})
+
+	When("a jobset is created with DNS hostnames enabled", func() {
+		It("should create all jobs and headless services, then complete successfully once all jobs are completed", func() {
+			By("creating a new JobSet")
+			ctx := context.Background()
+
+			// Create JobSet.
+			js := constructJobSet(&jobSetArgs{
+				name:               "js-enable-hostnames",
+				namespace:          ns.Name,
+				numJobs:            3,
+				enableDNSHostnames: true,
+			})
+			Expect(k8sClient.Create(ctx, js)).Should(Succeed())
+
+			// We'll need to retry getting this newly created JobSet, given that creation may not immediately happen.
+			By("checking JobSet was created successfully")
+			Eventually(jobSetExists, timeout, interval).WithArguments(js).Should(BeTrue())
+
+			By("checking JobSet eventually has 3 active jobs")
+			var childJobList batchv1.JobList
+			Eventually(func() (int, error) {
+				if err := k8sClient.List(ctx, &childJobList, client.InNamespace(js.Namespace)); err != nil {
+					return -1, err
+				}
+				return len(childJobList.Items), nil
+			}, timeout, interval).Should(Equal(3))
+
+			By("checking JobSet eventually has 3 headless services")
+			Eventually(func() (int, error) {
+				var svcList corev1.ServiceList
+				if err := k8sClient.List(ctx, &svcList, client.InNamespace(js.Namespace)); err != nil {
+					return -1, err
+				}
+				return len(svcList.Items), nil
+			}).Should(Equal(3))
+
+			By("checking JobSet status is completed once all its jobs are completed")
+			// Mark jobs as complete.
+			for _, job := range childJobList.Items {
+				job.Status.Conditions = append(job.Status.Conditions, batchv1.JobCondition{
+					Type:   batchv1.JobComplete,
+					Status: corev1.ConditionTrue,
+				})
+				Expect(k8sClient.Status().Update(ctx, &job)).Should(Succeed())
+			}
+			// Check JobSet has completed.
+			Eventually(assertJobSetStatus, timeout, interval).WithArguments(js, jobset.JobSetCompleted).Should(Equal(true))
+		})
+	})
+})
+
+func jobSetExists(js *jobset.JobSet) bool {
+	err := k8sClient.Get(ctx, types.NamespacedName{Name: js.Name, Namespace: js.Namespace}, &jobset.JobSet{})
+	if err != nil {
+		return false
+	}
+	return true
+}
+
+func assertJobSetStatus(js *jobset.JobSet, condition jobset.JobSetConditionType) (bool, error) {
+	var fetchedJS jobset.JobSet
+	if err := k8sClient.Get(ctx, types.NamespacedName{Namespace: js.Namespace, Name: js.Name}, &fetchedJS); err != nil {
+		return false, err
+	}
+	for _, c := range fetchedJS.Status.Conditions {
+		if c.Type == string(condition) {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func constructJobSet(args *jobSetArgs) *jobset.JobSet {
```

Comment:
lets create a testing pkg that includes wrappers; see example here: https://github.com/kubernetes-sigs/kueue/blob/main/pkg/util/testing/wrappers.go

Reply:
Ok I created the testing package with wrappers for JobSet and ReplicatedJob, under /pkg/util/testing.